nengo_gui/gui.py

```diff
@@ -1,111 +1,103 @@
-                                    'default.py')
-        if filename is not None:
-            try:
-                filename = os.path.relpath(filename)
-            except ValueError:
-                # happens on Windows if filename is on a different
-                # drive than the current directory
-                filename = filename
-
-        if model is None and locals is not None:
-            model = locals.get('model', None)
-
-        self.filename = filename
-        self.model = model
-        self.locals = locals
-        self.allow_file_change = allow_file_change
-        self.filename_cfg = cfg
-        self.interactive = interactive
-        self.backend = backend
-
-    def start(self, port=8080, browser=True, password=None):
-        """Start the web server"""
-        print("Starting nengo server at http://localhost:%d" % port)
-        if password is not None:
-            nengo_gui.server.Server.add_user('', password)
-            addr = ''
-        else:
-            addr = 'localhost'
-        nengo_gui.server.Server.start(self, port=port, browser=browser,
-                                      addr=addr, interactive=self.interactive)
+        self.server = GuiServer(
+            self.model_context, server_settings, page_settings)
 
-    def prepare_server(self, port=8080, browser=True):
-        return nengo_gui.server.Server.prepare_server(
-            self, port=port, browser=browser)
+    def start(self):
+        """Start the backend server and wait until it shuts down."""
+        try:
+            self.server.serve_forever(poll_interval=0.02)
+        except ServerShutdown:
+            self.server.shutdown()
+        finally:
+            self.server.wait_for_shutdown(0.05)
 
-    def begin_lifecycle(self, server):
-        nengo_gui.server.Server.begin_lifecycle(
-            server, interactive=self.interactive)
 
-    def create_page(self, filename, reset_cfg=False):
-        """Create a new Page with this configuration"""
-        page = nengo_gui.page.Page(self, filename=filename, reset_cfg=reset_cfg)
-        self.pages.append(page)
-        return page
+class InteractiveGUI(GUI):
+    """Creates a nengo_gui backend server and provides some useful information
+    on stdout. Also registers handlers to allow a server shutdown with Ctrl-C.
 
-    def remove_page(self, page):
-        self.pages.remove(page)
+    Parameters
+    ----------
+    model_context : nengo_gui.backend.backend.ModelContext
+        Model and its context served by the backend.
+    page_settings : nengo_gui.page.PageSettings, optional
+        Frontend page settings.
+    port : int
+        Port to listen on.
+    password : str, optional
+        Password required to connect to the backend.
+    """
+
+    def start(self):
+        protocol = 'https:' if self.server.settings.use_ssl else 'http:'
+        print("Starting nengo server at %s//%s:%d" %
+              (protocol, self.server.server_name, self.server.server_port))
```

Comment:
And, I think we should drop support for linux ... so now all the 
problems are fixed!

tcstewar wrote:
>
> In nengo_gui/gui.py
> <https://github.com/nengo/nengo_gui/pull/673#discussion_r65218741>:
>
> > + Parameters
> > + ----------
> > + model_context : nengo_gui.backend.backend.ModelContext
> > + Model and its context served by the backend.
> > + page_settings : nengo_gui.page.PageSettings, optional
> > + Frontend page settings.
> > + port : int
> > + Port to listen on.
> > + password : str, optional
> > + Password required to connect to the backend.
> > + """
> > +
> > + def start(self):
> > + protocol = 'https:' if self.server.settings.use_ssl else 'http:'
> > + print("Starting nengo server at %s//%s:%d" %
> > + (protocol, self.server.server_name, self.server.server_port))
>
> Sure, as long as I can drop support for OS X as well.... ;)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nengo/nengo_gui/pull/673/files/c2ccf0beb2810afe0d282638f57839261ad8c5ae#r65218741>, 
>
> or mute the thread
> <https://github.com/notifications/unsubscribe/AB5JU0M1_b3EudqSJk61jrYcgvFYS4IEks5qHGRigaJpZM4HVAVk>.